### ISSUE TITLE
Unable to write doc comment or metadata on primary constructor

### ISSUE THREAD

**Issue by srawlins** [dart-lang/language#3300]
I believe that neither the [spec for extension types](https://github.com/dart-lang/language/blob/main/accepted/future-releases/extension-types/feature-specification.md) nor the [proposed spec for primary constructors](https://github.com/dart-lang/language/pull/3023) allow for doc comments or metadata (annotations) on a primary constructor.

Maybe that's fine. Let's examine each of these cases:

#### Cannot specify doc comment on an extension type's primary constructor

Well, an extension type's primary constructor can only have one parameter. The meaning of that parameter is always the same. Maybe it is difficult to conceive of a case where a doc comment specific to the primary constructor is important.

If the ability to specify doc comments is important, maybe we can just introduce a [doc comment directive](https://github.com/dart-lang/dartdoc/wiki/Doc-comment-directives), e.g.

```dart
/// An extension type.
///
/// {@primary-constructor}
/// Wraps an [int] as a [V1].
/// {@end-primary-constructor}
extension type V1(int it) {
```

(Doc comment directives use kebab case; it is important to use as many cases in a programming language as possible.)

#### Cannot specify metadata on an extension type's primary constructor

This one worries me because I have to imagine there are plenty of cases where users desire to annotate a constructor. Maybe many cases can be handled by just assuming an annotation on the extension type is intended for the primary constructor. Maybe not.

I want to highlight the 3 primary purposes of annotations:

* trigger static analysis, like `@visibleForTesting`, `@experimental`, etc.
* trigger/inform code generation. I'd like to look at what code generation systems might use metadata on constructors, especially built_value, json_serializable, freezed.
* (future) trigger macro execution. The possibilities are pretty open wide here. CC @jakemac53

#### Cannot specify doc comment on a (class's, etc.) primary constructor

Move the constructor (and fields) to be written explicitly, and then document the constructor as usual. The same goes for documenting the fields.

#### Cannot specify metadata on a (class's, etc.) primary constructor

Move the constructor (and fields) to be written explicitly, and then annotate as usual.

**Comment by leafpetersen** [comment]
I'm not sure about the metadata, but the answer in Kotlin to the questions about the doc comments, is that all of that goes in the doc comment for the [class](https://kotlinlang.org/docs/kotlin-doc.html#kdoc-syntax).

**Comment by lrhn** [comment]
Putting the constructor documentation into the class documentation worried for Kotlin because they have doc-tags like `@param` and `@return`to say that the next paragraph applies to a specific part of the declaration.

Dart has so far chosen not to have doc tags, and rely on convention like a paragraph starting with "Returns" being the documentation of the returned value.

That means that our tools can't easily provide the documentation for _just_ a single parameter the way Java/Kotlin can.
And convention alone probably won't help us provide documentation for _just_ the constructor taken from the class documentation.

We can document the individual parameters of we want to, as 
```dart
class Foo(
  /// The number of bars.
  int bar,
  /// ...
) ...
```
But there is no good place to put documentation for the constructor.

Maybe if we put the constructor last:
```dart
class Foo implements FooBar
    /// Constructs a foo.
    /// 
    /// The `bar` is the number of bars.
    new(
       /// Number of bars.
       final int bar,
    ) {
  .. 
} 
```
Where you can document the constructor by putting a comment before the primary constructor declaration, and the fields out introduces by putting comments on the field-declaring parameters.

**Comment by eernstg** [comment]
> .. there is no good place to put documentation for the constructor.

True, but this might be OK. Primary constructors (as I've been thinking about them while writing [this proposal](https://github.com/dart-lang/language/pull/3023), and I think I'm not alone in that respect) are optimized for simple and small declarations. This might justify the advice that the documentation of the constructor is integrated into the documentation of the class, and DartDoc pages would show the documentation for the class as the documentation of the primary constructor as well. In that case it isn't necessary to introduce tags like `@primaryConstructor`.

If there is a need to add more detail then the primary constructor in the header can be turned into a primary constructor in the body:

```dart
/// An extension type.
extension type V1 {
  /// Wraps an [int] as a [V1].
  primary V1(int it);
  ...
}
```

**Comment by lrhn** [comment]
> To me it's counter intuitive that we don't allow:
> ```dart
> class Foo {
>   Foo({
>     // Allow the IDE to still show docs
>     /// Documentation for a
>     int? a,
>   }) : _a = a;
> 
>   final int? a;
> }
> ```

**Comment by rrousselGit** [comment]
To be clear, I wasn't necessarily talking about private parameters here but rather initializers in general.
Flutter suffers from this quite a bit. Lots of widgets defer rely on initializers. Private fields on only one case. For example I've also seen:

```dart
Class({required List list}):
  list = list.map(...).toList();
```

or:

```dart
Class({Color? primaryColor, this.scheme}):
  primaryColor = primaryColor ?? scheme?[0] ?? fallback;
```

This makes dartdoc a bit unpredictable. Sometimes the IDE shows useful doc; sometimes it doesn't.

And it can vary between two versions of the same package. Because after a refactoring, maybe an initializer was added/removed.

**Comment by goderbauer** [comment]
I've been thinking a little bit about how primary constructors could help with cutting down verbosity in Flutter - and I share some of the concerns w.r.t. to documentation discussed in this issue:

In Flutter, the doc comments on constructors are mostly just there to check the box that all public members have docs. They rarely add anything useful. Therefore, I don't think we need a dedicated place for docs of the primary constructor itself. Usually, the class comment as a whole is just sufficient here.

However, we do need a place to provide docs for the individual parameters of the primary constructor (the docs that traditionally would just go on the fields). There's lots of great documentation on the fields today that I wouldn't want to loose and I also wouldn't want to turn the class docs into a wall of text that include documentation for each argument. Something with a little more structure would be desirable.

I really like putting the parameter docs inline into the primary constructor:

```dart
/// A widget that shows the current count along with a button to increase it.
class Counter const (
  super.key, 
  
  /// The name of the counter.
  ///
  /// Will be displayed in the UI.
  String name, 
  
  /// The starting value for the counter.
  ///
  /// Defaults to 0.
  int startValue = 0,
) extends StatelessWidget {
  // ...
}
```
